### PR TITLE
Enable conda runtime creation in workers on windows

### PR DESCRIPTION
**Pull request by mattip**
## Why are these changes needed?
Continuation of #22449 

Fix the way a conda environment is created for a local worker to reuse the ray installation. The base problem was that `site.getsitepackages()[0]` did not return the site-packages library on windows. So I changed the call to use the more accepted `sysconfig.get_paths()['purelib']`. 

Along the way, I also made the test pass when using a `pip install -e .` which does not actually install ray into the site-packages directory, but uses a shim.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mattip** [comment]
@architkulkarni

File: python/ray/_private/runtime_env/conda.py
Comment (mattip) [inline python/ray/_private/runtime_env/conda.py:101]:
```suggestion
            # a `pip install -e .` hits this path
```

File: python/ray/_private/runtime_env/context.py
Comment (mattip) [inline python/ray/_private/runtime_env/context.py:56]: moving the logging to _before_ the command so that the message will be emitted even if the command fails.

**Comment by architkulkarni** [comment]
This looks reasonable, I'm not super familiar how the site mechanism works in Python so I think it would be helpful to have a review from @simon-mo.  @simon-mo would you mind taking a look?

Also @mattip, do you know if the Windows tests are currently being run with the `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` flag, or if they're using the wheel built in CI like in the Mac/Linux tests (this part https://github.com/ray-project/ray/blob/77c4c1e48e02c90660fd74b73ac694786b198141/python/ray/_private/runtime_env/conda.py#L119-L125)?  Using the wheel is a bit more faithful to the codepath used by the released versions of Ray, but either way is fine.

**Comment by mattip** [comment]
I will check that `RAY_CI_POST_WHEEL_TESTS`. I think we get a hint since this passed locally but failed on CI around the use of a `whl`:

The windows run of the enabled test is getting [an error](https://buildkite.com/ray-project/ray-builders-pr/builds/28514#24ddc0ec-9337-4103-81f7-07873e041511/2992-7331)
```
RuntimeError: Failed to install conda environment \
   C:\Users\ContainerAdministrator\AppData\Local\Temp\ray\session_2022-03-30_22-34-48_602017_6148\runtime_resources\conda\aa958e4ea4f7080354413b6d2cb9214637aa7044:
Installing pip dependencies: ...working... Ran pip subprocess with arguments:
['C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\ray\\session_2022-03-30_22-34-48_602017_6148\\runtime_resources\\conda\\aa958e4ea4f7080354413b6d2cb9214637aa7044\\python.exe', \
    '-m', 'pip', 'install', '-U', '-r', \
    'C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\ray\\session_2022-03-30_22-34-48_602017_6148\\runtime_resources\\conda\\condaenv.bixiqmor.requirements.txt'\
]
Processing c:\install\ray\.whl\ray-2.0.0.dev0-cp38-cp38-win_amd64.whl
...
ERROR: Could not install packages due to an OSError: [Errno 2] \
No such file or directory: 'C:\\Install\\ray\\.whl\\ray-2.0.0.dev0-cp38-cp38-win_amd64.whl'
```

**Comment by mattip** [comment]
On windows, the `WINDOWS_WHEELS` environment variable is not set during the `windows: Build and Test` CI step, so no wheel is built. This makes the conda test fail since `RAY_CI_POST_WHEEL_TESTS` is set. I am afraid setting the `WINDOWS_WHEELS` environment variable would break CI, I unset the `RAY_CI_POST_WHEEL_TESTS` option for windows. Let me know if this is unacceptable and we can try enabling `WINDOWS_WHEELS` for the whole test suite instead.

**Comment by mattip** [comment]
Hmm. Stuck between a rock and a hard place. Enabling `BUILD_WHEELS` (commit e998727) breaks the build: CI tries to build ray again after already building it once and fails. But without `BUILD_WHEELS`, with either `RAY_CI_POST_WHEEL_TESTS` disabled (commit 5c0469d59) or enabled like the original commit fail since they expect a wheel where there is none.

**Comment by architkulkarni** [comment]
> Hmm. Stuck between a rock and a hard place. Enabling `BUILD_WHEELS` (commit [e998727](https://github.com/ray-project/ray/commit/e998727f28c7b21820739c3b611bb7997ab92ab4)) breaks the build: CI tries to build ray again after already building ~it~ ray once and fails. But without `BUILD_WHEELS`, with either `RAY_CI_POST_WHEEL_TESTS` disabled (commit [5c0469d](https://github.com/ray-project/ray/commit/5c0469d5936a2e27e118d7f2da1d8200c215f876)) or enabled like the original commit fail since they expect a wheel where there is none.

I see... so the options are 
1. Somehow build the wheel in the Windows pipeline and make it appear in the right location, and enable `RAY_CI_POST_WHEEL_TESTS`.  
2. Don't build a wheel, and instead ensure `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` is used in CI so that runtime_env doesn't look for a wheel.

You said 1 isn't working because the second build fails.

I think 2 should work though.  I think the reason it isn't working is because we have `CI=1` set here https://github.com/ray-project/ray/commit/5c0469d5936a2e27e118d7f2da1d8200c215f876#diff-806d108e290c50e29ba31957e91323241f074d4f9a6c7f899ff74384655db9eeL183, but we also have https://github.com/ray-project/ray/blob/e998727f28c7b21820739c3b611bb7997ab92ab4/python/ray/tests/test_runtime_env_complicated.py#L34-L37.  

I think we just need to edit one of the parts linked above so that `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` is set in Windows CI, @mattip does that make sense?

**Comment by architkulkarni** [comment]
> > @architkulkarni or @mattip would you mind testing manually on linux/macos for this flag? If it works then this change LGTM
> 
> Sounds good, I'll check locally that the flag works on Mac and reply here if it works or not.

On master, `test_runtime_env_complicated` passes on my Mac laptop (it uses the `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` flag.). 

But on this PR, I get the following traceback:
```
――――――――――― ERROR at setup of test_client_tasks_and_actors_inherit_from_driver[ray start --head --ray-client-server-port 24001 --port 0] ――――――――――――

request = <FixtureRequest for <Function test_client_tasks_and_actors_inherit_from_driver[ray start --head --ray-client-server-port 24001 --port 0]>>

    def fill(request):
        item = request._pyfuncitem
        fixturenames = getattr(item, "fixturenames", None)
        if fixturenames is None:
            fixturenames = request.fixturenames
    
        if hasattr(item, 'callspec'):
            for param, val in sorted_by_dependency(item.callspec.params, fixturenames):
                if val is not None and is_lazy_fixture(val):
                    item.callspec.params[param] = request.getfixturevalue(val.name)
                elif param not in item.funcargs:
>                   item.funcargs[param] = request.getfixturevalue(param)

../../../../anaconda3/envs/ray-py38/lib/python3.8/site-packages/pytest_lazyfixture.py:37: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
conftest.py:275: in call_ray_start
    subprocess.check_output(command_args, stderr=subprocess.STDOUT)
../../../../anaconda3/envs/ray-py38/lib/python3.8/subprocess.py:415: in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

input = None, capture_output = False, timeout = None, check = True
popenargs = (['ray', 'start', '--head', '--ray-client-server-port', '24001', '--port', ...],), kwargs = {'stderr': -2, 'stdout': -1}
process = <subprocess.Popen object at 0x7fd0312faf40>
stdout = b'Traceback (most recent call last):\n  File "/Users/archit/anaconda3/envs/ray-py38/bin/ray", line 33, in <module>\n  ...te/usage/usage_lib.py", line 57, in <module>\n    import requests\nModuleNotFoundError: No module named \'requests\'\n'
stderr = None, retcode = 1

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '['ray', 'start', '--head', '--ray-client-server-port', '24001', '--port', '0']' returned non-zero exit status 1.

../../../../anaconda3/envs/ray-py38/lib/python3.8/subprocess.py:516: CalledProcessError
---------------------------------------------------------------- Captured log setup -----------------------------------------------------------------
WARNING  ray._private.runtime_env.conda:conda.py:86 Injecting /Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages to environment site-packages /Users/archit/anaconda3/envs/package-2.2.0/lib/python3.8/site-packages because _inject_current_ray flag is on.
WARNING  ray._private.runtime_env.conda:conda.py:86 Injecting /Users/archit/anaconda3/envs/ray-py38/lib/python3.8/site-packages to environment site-packages /Users/archit/anaconda3/envs/package-2.3.0/lib/python3.8/site-packages because _inject_current_ray flag is on.
                                                                                                                                        4% ▍         2022-04-04 13:12:50,843 ERROR services.py:1478 -- Failed to start the dashboard: Failed to start the dashboard, return code 1
Failed to read dashboard log: [Errno 2] No such file or directory: '/tmp/ray/session_2022-04-04_13-12-49_020360_60826/logs/dashboard.log'
2022-04-04 13:12:50,843 ERROR services.py:1479 -- Failed to start the dashboard, return code 1
Failed to read dashboard log: [Errno 2] No such file or directory: '/tmp/ray/session_2022-04-04_13-12-49_020360_60826/logs/dashboard.log'
Traceback (most recent call last):
  File "/Users/archit/ray/python/ray/_private/services.py", line 1441, in start_dashboard
    with open(dashboard_log, "rb") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/ray/session_2022-04-04_13-12-49_020360_60826/logs/dashboard.log'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/archit/ray/python/ray/_private/services.py", line 1452, in start_dashboard
    raise Exception(err_msg + f"\nFailed to read dashboard log: {e}")
Exception: Failed to start the dashboard, return code 1
Failed to read dashboard log: [Errno 2] No such file or directory: '/tmp/ray/session_2022-04-04_13-12-49_020360_60826/logs/dashboard.log'
[2022-04-04 13:12:53,091 E 60826 17594198] core_worker.cc:137: Failed to register worker 01000000ffffffffffffffffffffffffffffffffffffffffffffffff to Raylet. IOError: [RayletClient] Unable to register worker with raylet. No such file or directory
```
After this, if I try to do `import ray` I get `ModuleNotFoundError: No module named 'requests'`, even though I had `requests` installed before running the first test.  It seems like this could be related to the change in this PR, since the fixture `conda_envs` uses helper functions that are modified in this PR.

**Comment by mattip** [comment]
> After this, if I try to do `import ray` I get `ModuleNotFoundError: No module named 'requests'`

This is confusing. The PR (and the tests) should only be modifying the runtime environment in `/tmp/ray/session_XXX`. Perhaps the `ray.pth` file (that is now `ray_conda.pth` is somehow showing up in one of the directories in `sys.path`?

**Comment by mattip** [comment]
> After this, if I try to do import ray I get ModuleNotFoundError: No module named 'requests'

Reverting the `ray_path` change in 757120f fixed this for me on linux. @architkulkarni could you try again on macOS?

**Comment by architkulkarni** [comment]
> > After this, if I try to do import ray I get ModuleNotFoundError: No module named 'requests'
> 
> Reverting the `ray_path` change in [757120f](https://github.com/ray-project/ray/commit/757120fbfa0f2f76d8b94f1dee6a5efaeadaed93) fixed this for me on linux. @architkulkarni could you try again on macOS?

Fixed for me as well!  `runtime_env_complicated.py` passes for me locally on Mac

File: python/ray/_private/runtime_env/context.py
Comment (mattip) [reply python/ray/_private/runtime_env/context.py:56]: Maybe this should be in a separate PR if it is causing unsolvable problems.

**Comment by mattip** [comment]
The linkcheck failure is fixed on HEAD. It seems the CI does not merge with master before running.

**Comment by mattip** [comment]
Did I break all those non-windows test runs?

**Comment by architkulkarni** [comment]
> Did I break all those non-windows test runs?

Looks like it's unrelated: https://flaky-tests.ray.io/

**Comment by mattip** [comment]
When I run the tests via bazel [0] the `test_wait_mixup` failure to start up reproduces. When I run the tests via pytest [1] (after setting the environment like in the bazel command) they pass.

<details><summary>0</summary>

```
bazel test --config=ci --build_tests_only --test_env=CI=1 \
    --test_env=RAY_RUNTIME_ENV_LOCAL_DEV_MODE=1 --test_output=streamed \
    -- //python/ray/tests:test_async
...
============================= test session starts =============================
platform win32 -- Python 3.8.10, pytest-5.4.3, py-1.10.0, pluggy-0.13.1 -- C:\temp\ray_dev\Scripts\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\Matti\AppData\Local\Temp\Bazel.runfiles_z3qcxrpm\runfiles\com_github_ray_project_ray
plugins: asyncio-0.15.1, lazy-fixture-0.6.3, rerunfailures-10.2, timeout-1.4.2
collecting ... collected 9 items

::test_simple PASSED                                                     [ 11%]
(17:37:15) [1 / 2] Testing //python/ray/tests:test_async; 27s local
::test_gather PASSED                                                     [ 22%]
::test_wait PASSED                                                       [ 33%]
::test_wait_timeout PASSED                                               [ 44%]
(17:37:50) [1 / 2] Testing //python/ray/tests:test_async; 63s local
::test_gather_mixup PASSED                                               [ 55%]
::test_wait_mixup FAILED                                                 [ 66%]
::test_garbage_collection[ray_start_regular_shared0] PASSED              [ 77%]
::test_concurrent_future PASSED                                          [ 88%]
(17:38:32) [1 / 2] Testing //python/ray/tests:test_async; 104s local
::test_concurrent_future_many PASSED                                     [100%]
```

</details>
<details><summary>1</summary>

```
>python -m pytest python\ray\tests\test_async.py -vv
====================================================== test session starts =======================================================
platform win32 -- Python 3.8.10, pytest-5.4.3, py-1.10.0, pluggy-0.13.1 -- C:\temp\ray_dev\Scripts\python.exe
cachedir: .pytest_cache
rootdir: C:\oss\ray-git\python
plugins: asyncio-0.15.1, lazy-fixture-0.6.3, rerunfailures-10.2, timeout-1.4.2
collected 9 items

python\ray\tests\test_async.py::test_simple PASSED                                                                          [ 11%]
python\ray\tests\test_async.py::test_gather PASSED                                                                          [ 22%]
python\ray\tests\test_async.py::test_wait PASSED                                                                            [ 33%]
python\ray\tests\test_async.py::test_wait_timeout PASSED                                                                    [ 44%]
python\ray\tests\test_async.py::test_gather_mixup PASSED                                                                    [ 55%]
python\ray\tests\test_async.py::test_wait_mixup PASSED                                                                      [ 66%]
python\ray\tests\test_async.py::test_garbage_collection[ray_start_regular_shared0] PASSED                                   [ 77%]
python\ray\tests\test_async.py::test_concurrent_future PASSED                                                               [ 88%]
python\ray\tests\test_async.py::test_concurrent_future_many PASSED                                                          [100%]
```

**Comment by mattip** [comment]
Apparently the use of RAY_RUNTIME_ENV_LOCAL_DEV_MODE=1 causes many tests to fail on windows.

**Comment by mattip** [comment]
@architkulkarni using RAY_RUNTIME_ENV_LOCAL_DEV_MODE=1 is causing problems in many places:
- I needed to increase a timeout in 84add62
- many tests are timing out or failing to run properly.

The problem is that mode turns on `self["_inject_current_ray"] = True`. If I skip this, things start to work again. https://github.com/ray-project/ray/blob/52eaf020bcd4e8ebeb94af11a8039313a37488d1/python/ray/runtime_env.py#L390-L391

**Comment by architkulkarni** [comment]
> @architkulkarni using RAY_RUNTIME_ENV_LOCAL_DEV_MODE=1 is causing problems in many places:
> 
> * I needed to increase a timeout in [84add62](https://github.com/ray-project/ray/commit/84add625b082e76e4e8da97daed371be6210253b)
> * many tests are timing out or failing to run properly.
> 
> The problem is that mode turns on `self["_inject_current_ray"] = True`. If I skip this, things start to work again.
> 
> https://github.com/ray-project/ray/blob/52eaf020bcd4e8ebeb94af11a8039313a37488d1/python/ray/runtime_env.py#L390-L391

@mattip Interesting that things work after `inject_current_ray` is skipped... that's the only functionality of `RAY_RUNTIME_ENV_LOCAL_DEV_MODE`.  I would expect the runtime_env conda tests to fail, since Ray won't have been injected in the installed conda environment (so `ray` won't be importable by the workers.)

Would it be simpler to try option 1 from https://github.com/ray-project/ray/pull/23613#issuecomment-1087846395?  As long as the wheel is present in the right place on the CI machine, the tests should work just as they do for Mac/Linux.

**Comment by mattip** [comment]
> Would it be simpler to try option 1

Maybe, although I am not optimistic that installing a ray wheel into a conda environment will be fast on windows. File IO is one of the major slowdowns between windows and linux, so we may find ourselves in the same situation. As far as I can tell, RAY_RUNTIME_ENV_LOCAL_DEV_MODE triggers creation of a `pth` file. Or is there more file copying going on that would exlain the slowdown?

**Comment by architkulkarni** [comment]
Hmm, right, @simon-mo can correct me if I'm wrong but besides creating the `.pth` file the only thing that it's doing is copying all of Ray's pip dependencies into the conda `environment.yaml` spec: https://github.com/ray-project/ray/blob/77c4c1e48e02c90660fd74b73ac694786b198141/python/ray/_private/runtime_env/conda.py#L240

I suppose installing these could be slow, perhaps that could explain the slowdown?  I'm not really sure though, here's the list of dependencies:
```
    setup_spec.install_requires = [
        "attrs",
        "click >= 7.0, <= 8.0.4",
        "dataclasses; python_version < '3.7'",
        "filelock",
        "grpcio >= 1.28.1, <= 1.43.0",
        "jsonschema",
        "msgpack >= 1.0.0, < 2.0.0",
        "numpy >= 1.16; python_version < '3.9'",
        "numpy >= 1.19.3; python_version >= '3.9'",
        "protobuf >= 3.15.3",
        "pyyaml",
        "aiosignal",
        "frozenlist",
        "requests",
        "virtualenv",  # For pip runtime env.
    ]
```

**Comment by mattip** [comment]
Rebased and restarted this from the first commit. In discussion, we decided to try to simplify this to a single passing conda test, and then iterate in future PRs to get more conda testing to pass

**Comment by mattip** [comment]
@architkulkarni can you suggest where the simplest conda test is that is now skipped and should be able to pass? 

It seems the testing consistently turns off `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` and then has a `skipif(os.environ.get("CI") and sys.platform != "linux")` on all the relevant tests. For this to work, we would need to turn on `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` for this simple conda test on windows, and unskip it for windows.

**Comment by architkulkarni** [comment]
> @architkulkarni can you suggest where the simplest conda test is that is now skipped and should be able to pass?

Ah good question, I'd say `test_runtime_env_complicated.py::test_conda_create_task` is the simplest, followed closely by `test_conda_create_job_config` and `test_conda_create_ray_client` in the same file.

> It seems the testing consistently turns off `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` and then has a `skipif(os.environ.get("CI") and sys.platform != "linux")` on all the relevant tests. For this to work, we would need to turn on `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` for this simple conda test on windows, and unskip it for windows.

Yeah, I think there's two options:
1. Turn on `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` for Windows as you said, so it doesn't use a wheel, or
2. Ensure the test is running in the Windows pipeline with a wheel already built, and ensure `RAY_CI_POST_WHEEL_TESTS` is set and `RAY_RUNTIME_ENV_LOCAL_DEV_MODE` is unset so that it uses that wheel.  This is how it currently works in the Linux CI and is closest to how it works in the released Ray (it pulls the wheel from PyPI in that case.)

Either one is fine with me, you probably have the best sense of which one is easier.

**Comment by mattip** [comment]
I don't understand why the TestGC test in test_runtime_env_conda_and_pip is running (and failing), it should be skipped.

**Comment by mattip** [comment]
Good news: the test I enabled in [test_runtime_env_conda_and_pip_4](https://buildkite.com/ray-project/ray-builders-pr/builds/32661#540d7465-11b5-4001-9bc2-7ef54ce952da/3548-13543) seems to be passing. At least it is not failing.

**Comment by mattip** [comment]
One of the tests is crashing with 
```
[symbolize_win32.inc : 53] RAW: SymInitialize() failed: 2147483653
Fatal Python error: Aborted
```

this seems to be connected to  https://github.com/grpc/grpc/issues/24178 and may be fixed by grpcio1.46.2 which was released 4 days ago. The build is using grpcio==1.46.1, since 1.46.2 is not yet available on conda-forge.

**Comment by mattip** [comment]
I'm not sure what happened: it looks like CI ran again and now everything is passing?

**Comment by mattip** [comment]
Note that this has a fix for the agent ID, but the fix in #24968 is deeper and better. It would be nice to get that merged, then rebase this on top of it.

File: ci/ci.sh
Comment (mattip) [inline ci/ci.sh:186]: This seems to have solved a problem where the quoted `USERPROFILE` value somehow blocked setting `CI=1` and `RAY_CI_POST_WHEEL_TESTS=1`

**Comment by mattip** [comment]
Rebased on top of #24968, CI on windows is passing and I do not think the other failures are related.

File: ci/ci.sh
Comment (architkulkarni) [reply ci/ci.sh:186]: Interesting, not sure what's going on there but glad it worked!

File: python/ray/util/client/ray_client_helpers.py
Comment (architkulkarni) [inline python/ray/util/client/ray_client_helpers.py:42]: Nit: Should this be wrapped in an if statement so that it only happens on Windows?  Also, I wondered if this would slow things down excessively, but it looks like there are already some `sleep` statements above this line to account for Ray taking a few seconds to terminate.   Since this is just adding a few seconds more I think it's fine.

**Review by architkulkarni — APPROVED** [review]
This looks good!  Let me check the failures against the flaky test tracker.

File: ci/ci.sh
Comment (mattip) [reply ci/ci.sh:349]: I think it [adds a minute](https://buildkite.com/ray-project/ray-builders-pr/builds/33573#01810411-5b22-4f9b-97e8-f1f3a34c0056/2277-3357)

**Comment by architkulkarni** [comment]
- The "C++, Java, Libraries" failure is complaining about a [port collision,](https://buildkite.com/ray-project/ray-builders-pr/builds/33573#01810411-61f8-42cd-aa3b-b1958ab60ffe/5345-10628) it might be random? I didn't see it on the flaky test tracker.  It looks unrelated to this PR, so I restarted the test.

- "Apple Medium K-Z" It looks like the pytests all passed, but the job failed with "Error: The local post-command hook exited with status 1" and I couldn't find the root cause.  Restarting

- `test_client` is known flaky on the flaky test tracker
- `test_modin` is known flaky on the flaky test tracker

File: ci/ci.sh
Comment (architkulkarni) [reply ci/ci.sh:349]: Ah thanks a lot for the link!  That sounds fine to me, since the full job is ~120 minutes.

**Comment by architkulkarni** [comment]
Confirmed that "C++, Java and Libraries" failure is unrelated (it was broken on master), @simon-mo could you please merge this when you get a chance?

**Comment by mattip** [comment]
@architkulkarni the revert was reverted, this PR should be OK now.

**Comment by architkulkarni** [comment]
Great!  In that case @simon-mo could you please merge this when you get a chance?  Sorry for the back and forth

**Comment by DmitriGekhtman** [comment]
Poke @simon-mo

**Review by simon-mo — APPROVED** [review]

**Pull request merged**
